internal/permissions/permission_store.go

```diff
@@ -0,0 +1,143 @@
+package permissions
+
+import (
+	_ "embed"
+	"maps"
+	"slices"
+	"strings"
+
+	"github.com/bradenaw/juniper/xmaps"
+	"github.com/casbin/casbin/v2"
+	"github.com/casbin/casbin/v2/model"
+	"github.com/casbin/casbin/v2/persist"
+)
+
+// enum defining the possible actions a user has permission to do on an object.
+// an object is either a lexicon or a lexicon + record key
+type Action int
+
+const (
+	Read Action = iota
+	Write
+)
+
+var actionNames = map[Action]string{
+	Read:  "read",
+	Write: "write",
+}
+
+func (a Action) String() string {
+	return actionNames[a]
+}
+
+type Store interface {
+	HasPermission(
+		didstr string,
+		nsid string,
+		rkey string,
+		act Action,
+	) (bool, error)
+	AddLexiconReadPermission(
+		didstr string,
+		nsid string,
+	) error
+	RemoveLexiconReadPermission(
+		didstr string,
+		nsid string,
+	) error
+	ListReadPermissionsByLexicon() (map[string][]string, error)
+}
+
+type store struct {
+	enforcer *casbin.Enforcer
+	adapter  persist.Adapter
+}
+
+//go:embed model.conf
+var modelStr string
+
+func NewStore(adapter persist.Adapter, autoSave bool) (Store, error) {
+	m, err := model.NewModelFromString(modelStr)
+	if err != nil {
+		return nil, err
+	}
+	enforcer, err := casbin.NewEnforcer(m, adapter)
+	if err != nil {
+		return nil, err
+	}
+	// Auto-Save allows for single policy updates to take effect dynamically.
+	// https://casbin.org/docs/adapters/#autosave
+	enforcer.EnableAutoSave(autoSave)
+	return &store{
+		enforcer: enforcer,
+		adapter:  adapter,
+	}, nil
+}
+
+// HasPermission implements PermissionStore.
+func (p *store) HasPermission(
+	didstr string,
+	nsid string,
+	rkey string,
+	act Action,
+) (bool, error) {
+	return p.enforcer.Enforce(didstr, getObject(nsid, rkey), act.String())
+}
+
+// TODO: do some validation on input
```

Comment:
will we need to check for duplicates? Also is it possible to have conflicting permissions? don't need to figure out now but worth adding to the TODO

Reply:
Added an issue for this

Reply:
Added an issue